File: android/src/main/java/com/ammarahmed/rnadmob/nativeads/UnifiedNativeAdLoadedListener.java
package com.ammarahmed.rnadmob.nativeads;

import android.content.Context;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.WritableMap;
import com.google.android.gms.ads.nativead.NativeAd;
import com.google.android.gms.ads.nativead.NativeAd.OnNativeAdLoadedListener;
import com.google.android.gms.ads.nativead.NativeCustomFormatAd;

import java.lang.Long;
import java.util.List;

public class UnifiedNativeAdLoadedListener implements OnNativeAdLoadedListener, NativeCustomFormatAd.OnCustomFormatAdLoadedListener, NativeCustomFormatAd.OnCustomClickListener {
    String repo;
    List<RNAdMobUnifiedAdContainer> nativeAds;
    Context mContext;
    Integer totalAds;

    public UnifiedNativeAdLoadedListener(String repo, List<RNAdMobUnifiedAdContainer> nativeAds, Integer tAds, Context context) {
        this.repo = repo;
        this.nativeAds = nativeAds;
        this.mContext = context;
        this.totalAds = tAds;
    }

    @Override
    public void onNativeAdLoaded(NativeAd nativeAd) {
        if (this.nativeAds.size() > totalAds){
            // remove oldest ad if it is full
            RNAdMobUnifiedAdContainer toBeRemoved = null;
            Long time = System.currentTimeMillis();
            for (RNAdMobUnifiedAdContainer ad: this.nativeAds){
                if (ad.loadTime < time && ad.references <=0){
                    time = ad.loadTime;
                    toBeRemoved = ad;
                }
            }
            if (toBeRemoved !=  null){
                toBeRemoved.unifiedNativeAd.destroy();
                this.nativeAds.remove(toBeRemoved);
            }
        }
        this.nativeAds.add(new RNAdMobUnifiedAdContainer(nativeAd, System.currentTimeMillis(), 0));
        WritableMap args = Arguments.createMap();
        args.putInt(this.repo, this.nativeAds.size());
        EventEmitter.sendEvent((ReactContext) this.mContext, CacheManager.EVENT_AD_PRELOAD_LOADED, args);
    }

    @Override
    public void onCustomFormatAdLoaded(@NonNull NativeCustomFormatAd nativeCustomFormatAd) {
        if (this.nativeAds.size() > totalAds){
            // remove oldest ad if it is full
            RNAdMobUnifiedAdContainer toBeRemoved = null;
            Long time = System.currentTimeMillis();
            for (RNAdMobUnifiedAdContainer ad: this.nativeAds){
                if (ad.loadTime < time && ad.references <=0){
                    time = ad.loadTime;
                    toBeRemoved = ad;
                }
            }
            if (toBeRemoved !=  null){
                toBeRemoved.unifiedNativeAd.destroy();
                this.nativeAds.remove(toBeRemoved);
            }
        }
        this.nativeAds.add(new RNAdMobUnifiedAdContainer(nativeCustomFormatAd, System.currentTimeMillis(), 0));
        WritableMap args = Arguments.createMap();
        args.putInt(this.repo, this.nativeAds.size());
        EventEmitter.sendEvent((ReactContext) this.mContext, CacheManager.EVENT_AD_PRELOAD_LOADED, args);
    }


    @Override
    public void onCustomClick(@NonNull NativeCustomFormatAd nativeCustomFormatAd, @NonNull String s) {

    }
}

File: ios/RNAdMobManager/UnifiedNativeAdLoadedListener.m
//
//  OnUnifiedNativeAdLoadedListener.m
//  react-native-admob-native-ads
//
//  Created by Ali on 8/25/21.
//

#import <Foundation/Foundation.h>
#import "UnifiedNativeAdLoadedListener.h"
#import "RNAdMobUnifiedAdContainer.h"
#import "EventEmitter.h"
#import "CacheManager.h"

@implementation UnifiedNativeAdLoadedListener

-(instancetype) initWithRepo:(NSString *)repo nativeAds:(NSMutableArray<RNAdMobUnifiedAdContainer *> *) nativeAds tAds:(int)tAds {
    _repo = repo;
    _nativeAds = nativeAds;
    _totalAds = tAds;
    return self;
}

- (void)adLoader:(nonnull GADAdLoader *)adLoader didReceiveNativeAd:(nonnull GADNativeAd *)nativeAd {
    [self handleReceivedAd:nativeAd];
}

- (void)adLoader:(nonnull GADAdLoader *)adLoader didReceiveCustomNativeAd:(nonnull GADCustomNativeAd *)customNativeAd {
    [self handleReceivedCustomNativeAd:customNativeAd];
}

- (void)handleReceivedAd:(id)ad {
    long long time = (long long)([[NSDate date] timeIntervalSince1970] * 1000.0);
    if (self.nativeAds.count > _totalAds) {
        // remove oldest ad if it is full
        RNAdMobUnifiedAdContainer *toBeRemoved = nil;

        for (RNAdMobUnifiedAdContainer *adContainer in _nativeAds) {
            if (adContainer.loadTime < time && adContainer.references <= 0) {
                time = adContainer.loadTime;
                toBeRemoved = adContainer;
            }
        }
        if (toBeRemoved != nil) {
            toBeRemoved.unifiedNativeAd = nil;
            [self.nativeAds removeObject:toBeRemoved];
        }
    }
    RNAdMobUnifiedAdContainer *container = [[RNAdMobUnifiedAdContainer alloc] initWithAd:ad loadTime:time showCount:0];
    [self.nativeAds addObject:container];

    NSMutableDictionary* args = [[NSMutableDictionary alloc] init];
    [args setObject:[NSNumber numberWithInteger:_nativeAds.count] forKey:_repo];
    [EventEmitter.sharedInstance sendEvent:[CacheManager EVENT_AD_PRELOAD_LOADED:_repo] dict:args];
}

-(void)handleReceivedCustomNativeAd:(GADCustomNativeAd *)customNativeAd {
    long long time = (long long)([[NSDate date] timeIntervalSince1970] * 1000.0);
    if (self.nativeAds.count > _totalAds) {
        // remove oldest ad if it is full
        RNAdMobUnifiedAdContainer *toBeRemoved = nil;

        for (RNAdMobUnifiedAdContainer *adContainer in _nativeAds) {
            if (adContainer.loadTime < time && adContainer.references <= 0) {
                time = adContainer.loadTime;
                toBeRemoved = adContainer;
            }
        }
        if (toBeRemoved != nil) {
            toBeRemoved.unifiedNativeAd = nil;
            [self.nativeAds removeObject:toBeRemoved];
        }
    }
    RNAdMobUnifiedAdContainer *container = [[RNAdMobUnifiedAdContainer alloc] initWithCustomNativeAd:customNativeAd loadTime:time showCount:0];
    [self.nativeAds addObject:container];

    NSMutableDictionary* args = [[NSMutableDictionary alloc] init];
    [args setObject:[NSNumber numberWithInteger:_nativeAds.count] forKey:_repo];
    [EventEmitter.sharedInstance sendEvent:[CacheManager EVENT_AD_PRELOAD_LOADED:_repo] dict:args];
}

- (void)adLoader:(nonnull GADAdLoader *)adLoader didFailToReceiveAdWithError:(nonnull NSError *)error {

}

@end
